### DISCUSSION TITLE
custom fonts: better familly names

### DISCUSSION THREAD

**Discussion by frantisekhanzlikbl** [ryanoasis/nerd-fonts#1361] [Q&A] [answered]
I'm trying to patch a custom build of Iosevka, but I keep failing to get predictable family names for the resulting patched files.

The original unpatched files have family set to `iosevka polarised term`:
```console
$ otfinfo --info iosevka-polarised-term-regular.ttf
Family:              iosevka polarised term
Subfamily:           Regular
Full name:           iosevka polarised term
PostScript name:     iosevka-polarised-term
Preferred family:    iosevka polarised term
```

and as such, I'd expect the patched files to have their family set to the original family with something appended, e.g. `iosevka polarised term NF` or `iosevka polarised term Nerd Font`. instead, the patcher somehow magically generates `IosevkaPolarisedTerm Nerd Font`:
```console
$ otfinfo --info IosevkaPolarisedTermNerdFontPlusFontLogosPlusMaterialDesignIcons-Regular.ttf
Family:              IosevkaPolarisedTerm Nerd Font
Subfamily:           Regular
Full name:           IosevkaPolarisedTerm Nerd Font
PostScript name:     IosevkaPolarisedTermNFLMDI
Preferred family:    IosevkaPolarisedTerm Nerd Font Plus Font Logos Plus Material Design Icons
```

I have no idea where it took `IosevkaPolarisedTerm` from, but it doesn't appear to be anywhere in the original font's metadata. I tried using all `--makegroups` 0–6, none of which seem to help (with `0` producing even worse output of `iosevka Nerd Font`, losing the rest of the family name somewhere). The preferred family is also absolutely horrendous, because that's what fontconfig (and everything that takes font names from it) reports it as, making the output mostly unreadable.

Therefore, I would therefore like to ask:
1. is this behaviour expected? is there a good reason for transforming the original family name instead of only appending to it?
2. if so, can I disable it somehow?
3. as a workaround, can I somehow force a different family name?

Thanks in advance.

**Comment by Finii** [comment, marked as answer]
Lets see what I can answer

1a) Yes. The original name is always CamelCased and blanks removed. Then the "Nerd Font" stuff is appended. For example `JetBrains Mono` becomes `JetBrainsMono Nerd Font`.
1b) I do not know the reason for this.
2) There is `makegroups -1` which prevents any renaming. But then the original name is kept and nothing is appended.
3) Maybe `--name` is what you want?

Remarks:
`--makegroups 0` is just for people that want to have the years old buggy behaviour, which is deprecated and should not be used. The legacy code is still in and can be selected, but it will be removed somewhen (and then there is no `0` anymore).

**Reply by Finii** [reply]
`Nerd Font Plus Font Logos Plus Material Design Icons` ah, you do only a partial set patch. That long names are really ... interesting, but then, I have no good other idea. Maybe you have?

**Reply by frantisekhanzlikbl** [reply]
Thanks! That was exactly what I needed to know.

> The original name is always CamelCased and blanks removed.

Ah. All I really need is for the names to be predictable, as I need to know the resulting font names ahead of them being actually built. It's annoying to have to do the CamelCasing in my code too, but at least it's consistent.

> There is makegroups -1 which prevents any renaming.

I know about that flag, but I want to have both the patched and unpatched fonts installed at the same time, if, for whatever reason, some apps don't play along with the patched versions.

> Maybe --name is what you want?

That sounds like the perfect solution. Unfortunately, I can't find that flag anywhere in `font-patcher`'s `--help`, and supplying it manually also errors out. I'm on `Nerd Fonts Patcher v3.0.2 (4.4.0) (ff 20230101)`; is it perhaps something recently introduced?

> --makegroups 0 is just for people that want to have the years old buggy behaviour, which is deprecated and should not be used. The legacy code is still in and can be selected, but it will be removed somewhen (and then there is no 0 anymore).

I figured out as much, but gave it a try anyway. to see if the old generation perhaps worked better in this regard.

> ah, you do only a partial set patch.

Yeah, sorry, I should've mentioned that. I actually got rid of two glyph sets already, to make the names more manageable :smile: 

I'm guessing that the used sets are added to the family to be able to distinguish different variations better, but as it is now, they get unwieldy rather fast. I kinda like the few-letter indicators used in PostScript names: `IosevkaPolarisedTermNFLMDI`. perhaps those could be used for the preferred family too, with the full set names being appended to the description?

**Reply by Finii** [reply]
`--name` has been introduced with 4.5.0. You have 4.4.0.

The [docker thing](https://hub.docker.com/r/nerdfonts/patcher) and the [zip archive
](https://github.com/ryanoasis/nerd-fonts/blob/master/FontPatcher.zip) are always up to HEAD.
Current HEAD is `Nerd Fonts Patcher v3.0.2-62 (4.5.2) (ff 20230101)`

<img width="632" alt="Screenshot 2023-09-29 at 14 15 32" src="https://github.com/ryanoasis/nerd-fonts/assets/16012374/bd790c87-baba-44e5-9713-8a9d49c0ae18">

Thanks a lot for the idea on partial set naming!

**Reply by frantisekhanzlikbl** [reply]
Welp, that explains it. Turns out, my distro packages an old version of the patcher. I'll go fix that and implement the case transformation as a workaround for now. Thank you so much for your time!

Do you think it would be worth it to mention the transformation somewhere in the docs? If so, would you accept a PR? Which files should I mention it in? Also, should I update the "Full options:" section in readme to include the `--name` flag?

**Reply by Finii** [reply]
Well, `v4.4.0`is the patcher released with `Nerd Fonts v3.0.1`.

It is just that docker and zip get updated everytime something is pushed to the default branch, i.e. they follow HEAD commit by commit.

It is always good to improve the documentation :-) And of course a PR is welcome. But I have no clue where to mention it.

The full options section is updated on Nerd Font releases, which is not that often (the release date for `Nerd Fonts v3.1.0` is only some month overdue 😬 (*)

I'm not so sure the web page should show HEAD options, because a lot of people use the release version and strictly speaking that feature has not yet been released, so it should not be mentioned?

So usually I update the readme and the gh_pages only on (real) releases.
If someone has a HEAD patcher they should use `--help`.

Not really decided here, but leaning towards not-mentioning.

(*) There are some bugs I had hoped to fix but lacked the time to really solve, all of course Windows related ;-)

**Reply by frantisekhanzlikbl** [reply]
With some effort, I managed to use the `--name` flag on 35259124bd42527d179d671c2f1d5ec9d23f6ac5.

090fbbeebe256dc8078f2e3b3322fe804419b094 (master) kept erroring out for one of the iosevka files:
```python
  File "/nix/store/16s7v2fxdm3k8cq321bvivnldhyq0mr8-nerd-font-patcher-4.5.2/bin/.nerd-font-patcher-wrapped", line 2052, in <module>
    main()
  File "/nix/store/16s7v2fxdm3k8cq321bvivnldhyq0mr8-nerd-font-patcher-4.5.2/bin/.nerd-font-patcher-wrapped", line 2043, in main
    patcher.setup_font_names(f)
  File "/nix/store/16s7v2fxdm3k8cq321bvivnldhyq0mr8-nerd-font-patcher-4.5.2/bin/.nerd-font-patcher-wrapped", line 746, in setup_font_names
    n.rename_font(font)
  File "/nix/store/16s7v2fxdm3k8cq321bvivnldhyq0mr8-nerd-font-patcher-4.5.2/bin/../lib/name_parser/FontnameParser.py", line 301, in rename_font
    self.check_weights(font)
  File "/nix/store/16s7v2fxdm3k8cq321bvivnldhyq0mr8-nerd-font-patcher-4.5.2/bin/../lib/name_parser/FontnameParser.py", line 289, in check_weights
    if abs(os2_weight - ps_weight) > 50 or abs(os2_weight - name_weight) > 50:
TypeError: unsupported operand type(s) for -: 'int' and 'NoneType'
```
 
Unfortunately, even with the `--name` flag, the family name gets transformed to camel case (even though the spaces are kept). it does prevent the preferred family to be set, though, which is most of the way to what I need, so I'll stick with that.

Thanks again for the help. I'll put a note about the case transformation somewhere in the docs and submit a PR.

**Reply by Finii** [reply]
I will revisit the free-name-mechanism of `--name`. I just threw it in when implementing the name source selection stuff, without any concrete application and obviously not enough thought / energy.

Thanks for reporting the bug, will fix that also.

**Reply by Finii** [reply]
Maybe you can check #1364 with your iosevka variant? It should not error out anymore.
If not, maybe you can be more specific with the iosevka you used, so that I can reproduce and check the new code.
